File: modules/basic/ds/dataframe.cc
/** Copyright 2020-2023 Alibaba Group Holding Limited.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "basic/ds/dataframe.h"  // NOLINT(build/include)

#include <string>

#include "common/util/logging.h"  // IWYU pragma: keep

namespace vineyard {

class DataFrameBuilder;

DataFrameBuilder::DataFrameBuilder(Client& client)
    : DataFrameBaseBuilder(client) {}

const std::vector<json>& DataFrame::Columns() const { return this->columns_; }

std::shared_ptr<ITensor> DataFrame::Index() const {
  return values_.at("index_");
}

std::shared_ptr<ITensor> DataFrame::Column(json const& column) const {
  return values_.at(column);
}

const std::pair<size_t, size_t> DataFrame::partition_index() const {
  return std::make_pair(this->partition_index_row_,
                        this->partition_index_column_);
}

const std::pair<size_t, size_t> DataFrame::shape() const {
  if (values_.empty()) {
    return std::make_pair(0, 0);
  } else {
    return std::make_pair(values_.begin()->second->shape()[0], columns_.size());
  }
}

const std::shared_ptr<arrow::RecordBatch> DataFrame::AsBatch(bool copy) const {
  size_t num_columns = this->Columns().size();
  int64_t num_rows = 0;
  std::vector<std::shared_ptr<arrow::Array>> columns(num_columns);
  std::vector<std::shared_ptr<arrow::Field>> fields(num_columns);
  for (size_t i = 0; i < num_columns; ++i) {
    // cast to arrow::Array
    auto cname = this->Columns()[i];
    std::string field_name;
    if (cname.is_string()) {
      field_name = cname.get_ref<std::string const&>();
    } else {
      field_name = json_to_string(cname);
    }
    auto df_col = this->Column(cname);
    num_rows = df_col->shape()[0];

    if (auto tensor = std::dynamic_pointer_cast<Tensor<int32_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<uint32_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<int64_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<uint64_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor = std::dynamic_pointer_cast<Tensor<float>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<double>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<std::string>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<uint8_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    } else if (auto tensor =
                   std::dynamic_pointer_cast<Tensor<int8_t>>(df_col)) {
      num_rows = tensor->shape()[0];
    }

    std::vector<std::shared_ptr<arrow::Buffer>> buffer{
        nullptr /* null bitmap */};

    // process the second buffer for std::string type
    if (auto tensor = std::dynamic_pointer_cast<Tensor<std::string>>(df_col)) {
      std::shared_ptr<arrow::Buffer> copied_buffer;
      if (copy) {
        CHECK_ARROW_ERROR_AND_ASSIGN(
            copied_buffer,
            df_col->buffer()->CopySlice(0, df_col->auxiliary_buffer()->size()));
      } else {
        copied_buffer = df_col->buffer();
      }
      buffer.push_back(copied_buffer);
    }

    // process buffer
    {
      std::shared_ptr<arrow::Buffer> copied_buffer;
      if (copy) {
        CHECK_ARROW_ERROR_AND_ASSIGN(
            copied_buffer,
            df_col->buffer()->CopySlice(0, df_col->buffer()->size()));
      } else {
        copied_buffer = df_col->buffer();
      }
      buffer.push_back(copied_buffer);
    }

    columns[i] = arrow::MakeArray(arrow::ArrayData::Make(
        FromAnyType(df_col->value_type()), num_rows, buffer));

    std::shared_ptr<arrow::Scalar> sca;
    CHECK_ARROW_ERROR_AND_ASSIGN(sca, columns[i]->GetScalar(0));

    DLOG(INFO) << "at column" << i << " start element : " << sca->ToString()
               << " value type: " << df_col->value_type()
               << " meta data type name:" << df_col->meta().GetTypeName()
               << std::endl;

    fields[i] = std::make_shared<arrow::Field>(
        field_name, FromAnyType(df_col->value_type()));
  }
  return arrow::RecordBatch::Make(arrow::schema(fields), num_rows, columns);
}

const std::pair<size_t, size_t> DataFrameBuilder::partition_index() const {
  return std::make_pair(this->partition_index_row_,
                        this->partition_index_column_);
}

void DataFrameBuilder::set_partition_index(size_t partition_index_row,
                                           size_t partition_index_column) {
  this->set_partition_index_row_(partition_index_row);
  this->set_partition_index_column_(partition_index_column);
}

void DataFrameBuilder::set_row_batch_index(size_t row_batch_index) {
  this->set_row_batch_index_(row_batch_index);
}

void DataFrameBuilder::set_index(std::shared_ptr<ITensorBuilder> builder) {
  this->values_.emplace("index_", builder);
}

std::shared_ptr<ITensorBuilder> DataFrameBuilder::Column(
    json const& column) const {
  return values_.at(column);
}

void DataFrameBuilder::AddColumn(json const& column,
                                 std::shared_ptr<ITensorBuilder> builder) {
  this->columns_.emplace_back(column);
  this->values_.emplace(column, builder);
}

void DataFrameBuilder::DropColumn(json const& column) {
  // FIXME: how to ensure the removed builder got destroyed/aborted.
}

Status DataFrameBuilder::Build(Client& client) {
  this->set_columns_(columns_);
  for (auto const& kv : values_) {
    this->set_values_(
        kv.first,
        std::dynamic_pointer_cast<ObjectBuilder>(kv.second)->Seal(client));
  }
  return Status::OK();
}

void GlobalDataFrame::PostConstruct(const ObjectMeta& meta) {
  if (meta.HasKey("partition_shape_row_")) {
    meta.GetKeyValue("partition_shape_row_", this->partition_shape_row_);
  }
  if (meta.HasKey("partition_shape_column_")) {
    meta.GetKeyValue("partition_shape_column_", this->partition_shape_column_);
  }
}

const std::vector<std::shared_ptr<DataFrame>> GlobalDataFrame::LocalPartitions(
    Client& client) const {
  std::vector<std::shared_ptr<DataFrame>> local_chunks;
  for (auto iter = LocalBegin(); iter != LocalEnd(); iter.NextLocal()) {
    local_chunks.emplace_back(*iter);
  }
  return local_chunks;
}

const std::pair<size_t, size_t> GlobalDataFrame::partition_shape() const {
  return std::make_pair(this->partition_shape_row_,
                        this->partition_shape_column_);
}

const std::pair<size_t, size_t> GlobalDataFrameBuilder::partition_shape()
    const {
  return std::make_pair(this->partition_shape_row_,
                        this->partition_shape_column_);
}

void GlobalDataFrameBuilder::set_partition_shape(
    const size_t partition_shape_row, const size_t partition_shape_column) {
  this->partition_shape_row_ = partition_shape_row;
  this->partition_shape_column_ = partition_shape_column;
  this->AddKeyValue("partition_shape_row_", partition_shape_row_);
  this->AddKeyValue("partition_shape_column_", partition_shape_column_);
}

void GlobalDataFrameBuilder::AddPartition(const ObjectID partition_id) {
  this->AddMember(partition_id);
}

void GlobalDataFrameBuilder::AddPartitions(
    const std::vector<ObjectID>& partition_ids) {
  this->AddMembers(partition_ids);
}

}  // namespace vineyard

File: rust/vineyard/src/ds/dataframe.rs
// Copyright 2020-2023 Alibaba Group Holding Limited.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use arrow_array::{ArrayRef, RecordBatch};
use arrow_schema::{DataType, Field, Schema};
use itertools::izip;
use serde_json::Value;

use super::tensor::*;
use crate::client::*;

#[derive(Default)]
pub struct DataFrame {
    meta: ObjectMeta,
    names: Vec<String>,
    columns: Vec<Box<dyn Tensor>>,
}

impl_typename!(DataFrame, "vineyard::DataFrame");

impl Object for DataFrame {
    fn construct(&mut self, meta: ObjectMeta) -> Result<()> {
        vineyard_assert_typename(typename::<Self>(), meta.get_typename()?)?;
        let size = meta.get_usize("__values_-size")?;
        self.names = Vec::with_capacity(size);
        self.columns = Vec::with_capacity(size);
        for i in 0..size {
            let name = meta.get_value(&format!("__values_-key-{}", i))?;
            let name = match name {
                Value::String(name) => name,
                _ => name.to_string(),
            };
            self.names.push(name);
            let column = meta.get_member_untyped(&format!("__values_-value-{}", i))?;
            self.columns.push(downcast_to_tensor(column)?);
        }
        return Ok(());
    }
}

register_vineyard_object!(DataFrame);

impl DataFrame {
    pub fn new_boxed(meta: ObjectMeta) -> Result<Box<dyn Object>> {
        let mut object = Box::<Self>::default();
        object.construct(meta)?;
        Ok(object)
    }

    pub fn num_columns(&self) -> usize {
        self.columns.len()
    }

    pub fn names(&self) -> &[String] {
        &self.names
    }

    pub fn name(&self, index: usize) -> &str {
        &self.names[index]
    }

    pub fn columns(&self) -> &[Box<dyn Tensor>] {
        &self.columns
    }

    pub fn column(&self, index: usize) -> ArrayRef {
        self.columns[index].array()
    }

    pub fn recordbatch(&self) -> Result<RecordBatch> {
        let mut columns = Vec::with_capacity(self.columns.len());
        for column in &self.columns {
            columns.push(column.array());
        }
        let types: Vec<DataType> = columns
            .iter()
            .map(|column| column.data_type().clone())
            .collect();
        let batch = RecordBatch::try_new(
            Arc::new(Schema::new(
                izip!(self.names.clone(), types)
                    .map(|(name, datatype)| Field::new(name, datatype, true))
                    .collect::<Vec<Field>>(),
            )),
            columns,
        )?;
        return Ok(batch);
    }
}

pub struct DataFrameBuilder {
    sealed: bool,
    names: Vec<String>,
    columns: Vec<Box<dyn Object>>,
}

impl ObjectBuilder for DataFrameBuilder {
    fn sealed(&self) -> bool {
        self.sealed
    }

    fn set_sealed(&mut self, sealed: bool) {
        self.sealed = sealed;
    }
}

impl ObjectBase for DataFrameBuilder {
    fn build(&mut self, _client: &mut IPCClient) -> Result<()> {
        if self.sealed {
            return Ok(());
        }
        self.set_sealed(true);
        return Ok(());
    }

    fn seal(mut self, client: &mut IPCClient) -> Result<Box<dyn Object>> {
        self.build(client)?;
        let mut meta = ObjectMeta::new_from_typename(typename::<DataFrame>());
        meta.add_usize("__values_-size", self.names.len());
        meta.add_isize("partition_index_row_", -1);
        meta.add_isize("partition_index_column_", -1);
        meta.add_isize("row_batch_index_", -1);
        for (index, (name, column)) in self.names.iter().zip(self.columns).enumerate() {
            meta.add_value(
                &format!("__values_-key-{}", index),
                Value::String(name.into()),
            );
            meta.add_member(&format!("__values_-value-{}", index), column)?;
        }
        let metadata = client.create_metadata(&meta)?;
        return DataFrame::new_boxed(metadata);
    }
}

impl DataFrameBuilder {
    pub fn new(
        client: &mut IPCClient,
        names: Vec<String>,
        arrays: Vec<arrow_array::ArrayRef>,
    ) -> Result<Self> {
        let mut columns = Vec::with_capacity(arrays.len());
        for array in arrays {
            columns.push(build_tensor(client, array)?);
        }
        return Ok(DataFrameBuilder {
            sealed: false,
            names,
            columns: columns,
        });
    }

    pub fn new_from_columns(names: Vec<String>, columns: Vec<Box<dyn Object>>) -> Result<Self> {
        return Ok(DataFrameBuilder {
            sealed: false,
            names,
            columns,
        });
    }
}
